cratetorrent/src/disk/io/file.rs

```diff
@@ -166,4 +227,100 @@ impl TorrentFile {
+    }
+}
+
+#[cfg(target_os = "linux")]
+fn convert_nix_error(err: nix::Error) -> std::io::Error {
+    match err {
+        nix::Error::Sys(errno) => std::io::Error::from_raw_os_error(errno as i32),
+
+        // preadv/pwrite never return them, but better safe than sorry
+        nix::Error::InvalidPath |
+        nix::Error::InvalidUtf8 |
+        nix::Error::UnsupportedOperation => {
+            log::warn!("Unexpected nix::Error kind ({}), falling back to last_os_error", err);
+            std::io::Error::last_os_error()
+        },
+    }
+}
+
+// Here's where the most effective implementation of `write_at/read_at` is chosen.
+// The trick is, linux-specific APIs do not modify the "seek cursor", while Windows' and others' do.
+// This discrepancy doesn't matter because nothing in this crate relies on this cursor
+// (the read/write offset is always provided explicitly).
+//
+// FIXME: On the other hand, if `TorrentInfo.handle` will ever become exposed in our public API,
+// this "jumping cursor" can potentially be observed by the outside world. It shouldn't be
+// a problem either because, well, I fail to see why someone would want to interact with
+// the file in a way where the cursor's positioning matters. But, at least, this should be
+// clearly documented and the user must be warned to perform his own seeks if need for
+// unaccounted reads/writes presents itself...
+cfg_if! {
+    if #[cfg(any(unix, target_os = "redox", target_os = "vxworks", target_os = "hermit"))] {
+        use std::os::unix::fs::FileExt;
+
+        #[inline]
+        fn write_all_at(file: &File, buf: &[u8], offset: u64) -> Result<(), WriteError> {
+            file.write_all_at(buf, offset).map_err(Into::into)
+        }
+
+        #[inline]
+        fn read_exact_at(file: &File, buf: &mut [u8], offset: u64) -> Result<(), ReadError> {
+            file.read_exact_at(buf, offset)
+                .map_err(|e| match e.kind() {
+                    ErrorKind::UnexpectedEof => ReadError::MissingData,
+                    _ => e.into()
+                })
+
+        }
+    } else if #[cfg(windows)] {
+        use std::os::windows::fs::FileExt;
+
+        fn write_all_at(file: &File, mut buf: &[u8], mut offset: u64) -> Result<(), WriteError> {
+            while !buf.is_empty() {
+                let written = unwrap_write_res!(file.seek_write(buf, offset));
+                offset += written;
+                buf = &buf[written..];
+            }
+
+            Ok(())
+        }
+
+        fn read_exact_at(file: &File, mut buf: &mut [u8], mut offset: u64) -> Result<(), ReadError> {
+            while !buf.is_empty() {
+                let read = unwrap_read_res!(file.seek_read(buf, offset));
+                offset += read;
+                buf = &mut buf[read..];
+            }
+
+            Ok(())
+        }
+    } else {
+        fn write_all_at(file: &File, buf: &[u8], offset: u64) -> Result<(), WriteError> {
+            file.seek(SeekFrom(offset))?;
```

Comment:
There might be a potential problem here for a future optimization (and in the read equivalent below).

While currently files are write locked when written, I think we can actually parallelize writes: because each piece maps to a disjunct portion of a file, there would be no clashes among writes.

If different pieces were written to the same file at the same time, there would be multiple calls to seek which I believe would be a race condition.

I think it's ok to leave as is for now with a note in the `TorrentFile::write` doc cautioning against the above.

Reply:
I just realized that reads to the same file are already done concurrently, so concurrent seeks could already be an issue there?

Reply:
True. I'll be thinking what to do, probably ditching the "generic" version and using platform specific APIs for each of the supported platforms..